### PR TITLE
refactor: collect all post install scripts

### PR DESCRIPTION
**Pull request by dead-horse**
回退到 1.1.0 的版本基础上改，还是按照之前的逻辑，只是将 postinstall 单独收集起来

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -214,3 +239,33 @@ function* linkLatestVersion(pkg, storeDir) {
...
+      if (postinstallScript) {
+        options.console.log(chalk.yellow('[%s@%s] scripts.postinstall: %j at %s'),
+          pkg.name, pkg.version, postinstallScript, root);
+        yield utils.runScript(root, postinstallScript, options);
+      }
+    } catch (err) {
+      if (task.optional) {
+        return console.warn(chalk.red('[%s@%s] optional error: %s'), err.stack);
+      }
+      err.message = `post install error, please remove node_modules before retry!\n${err.message}`;
```
Comment (dead-horse) [inline lib/index.js]: post install 出错的时候可能都已经 link 完了，所以只能提示下次安装前清理

File: lib/install.js [hunk 2]
```diff
@@ -162,15 +146,12 @@ function* _install(parentDir, pkg, options) {
...
+    yield postinstall(realPkg, realPkgDir, pkg.optional, options);
   } catch (err) {
     // delete donefile when install error, make sure next time install won't success.
-    const donefile = path.join(realPkgDir, '.tnpminstall.done');
-    if (yield fs.exists(donefile)) {
-      options.console.error(chalk.red('[%s@%s] install error: %s, parentDir: %s, remove %s\nerror stack: %s'),
-        realPkg.name, realPkg.version, err, parentDir, donefile, err.stack);
-      yield fs.unlink(donefile);
-    }
+    yield utils.rimraf(realPkgDir);
```
Comment (dead-horse) [inline lib/install.js]: 还有非 npm 的模块，还是直接删掉这个目录好了，下次再重新安装

File: lib/index.js [hunk 3]
```diff
@@ -126,7 +126,9 @@ module.exports = function*(options) {
   }
   yield linkTasks;
 
-  if (installRoot) yield postinstall(rootPkg, options.root, options);
+  if (installRoot) yield postinstall(rootPkg, options.root, false, options);
+
+  yield runPostInstallTasks(options);
```
Comment (fengmk2) [inline lib/index.js]: runPostInstallTasks 运行完再运行 root 的 postinstall 吧？

File: lib/index.js [hunk 4]
```diff
@@ -214,3 +239,33 @@ function* linkLatestVersion(pkg, storeDir) {
   const relative = yield utils.forceSymlink(realDir, linkDir);
   debug('%s@%s link %s => %s', pkg.name, pkg.version, linkDir, relative);
 }
+
+function* runPostInstallTasks(options) {
+  if (options.postInstallTasks.length) {
+    options.console.log(chalk.yellow('excute post install scripts...'));
+  }
+  for (const task of options.postInstallTasks) {
```
Comment (fengmk2) [inline lib/index.js]: debug 打印一下 postInstallTasks

File: lib/index.js (hunk 3 above)
Comment (dead-horse) [inline lib/index.js]: 上面那句 postinstall 其实就是把他加到 `options.postInstallTaks` 末尾了

File: lib/index.js (hunk 4 above)
Comment (dead-horse) [inline lib/index.js]: 执行的时候都会在下面打印出来的

**Comment by kevincaradant** [comment]
I tried the npminstall from the 'refactor_await_all' branch and there are no error now :). But npminstall is very more slow with this patch than before. Old version : 1min to install.    New version:  9 min to install :/ even if before we had errors with some under dev-dependencies. Do you think that you can do a patch to have an install more quickly ? :). thank you in advance

**Comment by dead-horse** [comment]
@kevincaradant this patch will only collects all post install scripts and execute after all dependencies install completed, so it won't affect install performance, maybe it is cause by network?

and you can find `excute post install scripts...` in the install logs, before this line is install dependencies and after this line is execute post install scripts. and you can get speed in the last line.

image

**Comment by fengmk2** [comment]
I run again the benchmark, result is not much diff as before:

```
------- npminstall@v1.1.1 -------
All packages installed (438 packages installed from npm registry, use 13s, speed 889.14kB/s, json 500(2.8MB), tarball 8.75MB)

real	0m13.675s
user	0m9.526s
sys	0m4.893s
------- npminstall@v1.1.1 with cache -------
All packages installed (438 packages installed from npm registry, use 12s, speed 239.06kB/s, json 501(2.81MB), tarball 0B)

real	0m12.382s
user	0m8.605s
sys	0m4.217s
------- npminstall@v1.1.1 --no-cache -------
real	0m13.324s
user	0m9.171s
sys	0m4.350s
--------- pnpm@0.18.0 -----------
real	0m18.139s
user	0m12.754s
sys	0m4.720s
```

File: lib/index.js [hunk 5]
```diff
@@ -214,3 +239,33 @@ function* linkLatestVersion(pkg, storeDir) {
...
+    try {
+      if (installScript) {
+        options.console.log(chalk.yellow('[%s@%s] scripts.install: %j at %s'),
+          pkg.name, pkg.version, installScript, root);
+        yield utils.runScript(root, installScript, options);
+      }
+      if (postinstallScript) {
+        options.console.log(chalk.yellow('[%s@%s] scripts.postinstall: %j at %s'),
+          pkg.name, pkg.version, postinstallScript, root);
+        yield utils.runScript(root, postinstallScript, options);
```
Comment (fengmk2) [inline lib/index.js]: also log the postInstall run time?

Comment (dead-horse) [inline lib/index.js]: added

Comment (dead-horse) [inline lib/index.js]: image

**Comment by fengmk2** [comment]
+1

**Comment by fengmk2** [comment]
1.2.0

**Comment by kevincaradant** [comment]
OK yes my bad , since i have directly upgraded in 1.2.0, all works fine. Less 1 min to install instead of 13 min with the classic npm:D.

Very very great guys ;)

**Comment by fengmk2** [comment]
And all tests in your project pass?

**Pull request merged**